lib/id3c/api/routes.py

```diff
@@ -283,6 +266,71 @@ def get_identifier_set_uses(*, session):
 
     return jsonify([ use._asdict() for use in uses ])
 
+@api_v1.route("/warehouse/sample", methods = ['POST'])
+@content_types_accepted(["application/json"])
+@check_content_length
+@authenticated_datastore_session_required
+def post_sample(*, session):
+    """
+    Create/update a single sample.
+
+    POST /warehouse/sample to validate, then create new or update
+    existing sample that passes validation.
+
+    JSON object with status (inserted, updated, or failed validation)
+    of the sample is returned.
+    """
+
+    sample = request.get_json()
+
+    LOG.debug(f"Creating/updating sample «{sample}»")
+    
+    # using DraftXValidator(schema).validate(...) instead of jsonschema.validate(...) to
+    # return accurate error message for "anyOf" requirement
+    try:
+        Draft7Validator(schemas.POST_SAMPLE_SCHEMA, format_checker=draft7_format_checker).validate(sample)
+    except ValidationError as e:
+        return str(e), 400
+
+    # Convert dates in YYYY-MM-DD format to MM/DD/YYYY for consistency with records
+    # created through manifest ETL process
+    for key in sample:
+        if key.endswith('_date'):
+            try:
+                sample[key] = datetime.strptime(sample[key], '%Y-%m-%d').strftime('%m/%d/%Y')
+            except:
+                pass
+    
+    # calculate hash based on serialized sample record
+    digest = sha1(json.dumps(sample, sort_keys=True).encode()).hexdigest()
+
+    provenance = {
+        "source": request.referrer or request.remote_addr or "?",
+        "method": request.method,
+        "path": request.path,
+        "timestamp": f'{datetime.now():%Y-%m-%d %H:%M:%S%z}',
+        "sha1sum": digest
+    }
+    sample.update(_provenance=provenance)
+
+    # Transform racks data into array to match previously established format
+    racks = {key: value for key, value in sample.items() if key.startswith('rack_')}
+    if racks:
+        sample["racks"]=[]
+        for k, v in racks.items():
+            sample["racks"].append(sample.pop(k))
+    
+    # Transform aliquots data into array to match previously established format
+    aliquots = {key: value for key, value in sample.items() if key.startswith('aliquot_')}
+    if aliquots:
+        sample["aliquots"]=[]
+        for k, v in aliquots.items():
+            sample["aliquots"].append(sample.pop(k))
+
+    result = datastore.store_sample(session, sample)
+
+    return (jsonify(result), 200) if "sample" in result else (jsonify(result), 400)
```

Comment:
I wonder if we want to be a bit more nuanced here. We could return a 400 if we specifically see a user error like in validation, but we may want to fall back to a 500 in the more general case that we can't attribute to an issue with the user's request. https://restfulapi.net/http-status-codes/

Reply:
Good suggestion. In testing it I wasn't able to get it to fail with anything that wasn't covered by JSON Schema or barcode validation, but will update it to return a 500 code in those cases.

Reply:
Thanks for adding the 500 response code. In case some ideas are helpful, here's a couple options for generating other errors:
* You can cause a `401` by supplying invalid credentials (or I think by shutting off your local database too)
* You can cause a `500` by modifying the application to make it throw an unhandled exception. It can be as basic as adding `raise Exception("Unknown error")` in `post_sample()`, or more nuanced -- like if you wanted to simulate a database issue, you could modify some of the SQL to reference a column that doesn't exist.

Reply:
Ok, thanks!  I realized that the `upsert_sample` function actually throws an exception in some cases, so I ended up wrapping that in a try/catch and sending back the error message in the API response. This is already deployed, but let me know if you other error handling suggestions other than what I added in c00de8098929e0d8230528baac58631ed5eb7ce9 and I can try to add them in when I implement a sample GET endpoint.